rfcs/0000-lifecycle-crd.md

```diff
@@ -0,0 +1,125 @@
+
+**Problem:**
+
+The CNB lifecycle is provided to kpack as in image reference in a configMap similar to how the completion and prepare
+images are passed to the kpack controller. This means that every builder will always use the same exact lifecycle, with
+no ability to specify an alternate one that may support different buildpack api or platform api versions. Due to kpack
+support of windows, a custom lifecycle image with specific layers for the linux and windows binaries is required
+(with a label specifying the layer diff id for each os). Because of this, a cve present in the lifecycle that has been 
+patched upstream, requires any user to build the kpack specific lifecycle image themselves if the kpack provided one
+has not been bumped.
+
+**Outcome:**
+
+Users will be able to create a CR that references a lifecycle image. The file structure in the image must match 
+that of the [CNB Lifecycle image](https://hub.docker.com/r/buildpacksio/lifecycle/tags).
+
+```
+.
+└── cnb/
+    └── lifecycle/
+        ├── analyzer
+        ├── builder
+        ├── detector
+        ├── exporter
+        ├── launcher
+        ├── lifecycle.toml
+        └── restorer
+```
+
+They will be able to use this lifecycle CR in any builder. A lifecycle will be shipped out of the box in kpack and 
+users will not be required to provide a lifecycle reference in the builder if they do not wish to override the default.
+Users should see no difference in functionality compared to the existing kpack lifecycle if they do not choose to 
+provide their own lifecycle CR.
+
+Instead of creating custom multi-layer lifecycle images, users can create a lifecycle CR for their desired runtime (i.e. windows or arm64).
+During builder creation, kpack will validate that the referenced lifecycle CR has the same OS and architecture as the base image.
+
+**Actions to take:**
+
+A new ClusterLifecycle CRD will be created. It will be very similar in structure to the ClusterStack Resource. 
+We will create a cluster scoped version of this resource. In the future, we may decide to create a namespaced version
+as well, but that is not in scope for this RFC. The name ClusterLifecycle aligns with kpack's existing cluster scoped 
+resources.
+
+This is an example instance of the proposed CRD:
+```yaml
+apiVersion: kpack.io/v1alpha2
+kind: ClusterLifecycle
+metadata:
+  name: sample-cluster-lifecycle
+spec:
+  serviceAccountRef:
+    name: sample-sa
+    namespace: sample-namespace
+  image: buildpacksio/lifecycle@sha256:f4ce143ea6bbc6b5be5f4d151aba8214324adb93bbd7e3b1f43cd134ad450bf7
+```
+
+
+The existing lifecycle reconciler can be updated to reconcile this new CRD instead of ConfigMaps.
+
+A new field will be added to the ClusterBuilder/Builder resources allowing the user to provide a reference to a lifecycle CR.
+This field will be optional and will default to the highest semantic version that is supported by the buildpacks in the builder and the platform. 
```

Comment:
```suggestion
This field will be optional and will default to the highest semantic version that is supported by the platform. 
```
Buildpacks are configured by the operator, so I don't think we can know which Buildpack APIs will be required... right? In theory (assuming we think most buildpacks in the wild have upgraded to at least Buildpack API 0.7 - since lifecycle 0.18.x and higher drops support for earlier APIs) it should be the highest semantic version that supports the Platform APIs [here](https://github.com/buildpacks-community/kpack/blob/ae5eb15b8b9632fda50bafc4d2fad0f8c2c8a2a8/pkg/apis/build/v1alpha2/build_pod.go#L1098)...right? In practice, that is going to mean the latest available lifecycle (until we deprecate more Platform APIs).

Reply:
we do validate that any buildpack referenced in the order is using buildpack apis supported by the lifecycle when constructing a builder https://github.com/buildpacks-community/kpack/blob/ae5eb15b8b9632fda50bafc4d2fad0f8c2c8a2a8/pkg/cnb/builder_builder.go#L188-L204

Reply:
Ah, I see - but that validation happens after the ClusterLifecycle is defined, no? So in the case that the provided lifecycle doesn't support a buildpack in the builder we just fail, right? (Instead of selecting a different lifecycle?)

Reply:
I think the only check we'd want when the cluster lifecycle is reconciled is the platform api compatibility. The buildpack api compatibility will continue to happen at builder creation